tweak heuristic for sdpa selection based off of *data* (and a decision tree)

Chillee opened the pull request:
High level approach:
1. I generated a bunch of data comparing FlashAttention and Cutlass implementations (https://pastebin.com/pe0j3YeK) 
2. I trained a decision tree using standard train/val split methodology and hyperparameter sweeps (https://pastebin.com/fjYX1HjR).
2a. I did a bunch of feature augmentation to capture interactions between features.

The heuristic I ended up with is:
```
use_flash = seq_len / (num_heads * batch_size) > 6
```

TL;DR: On my dataset, where FlashAttention and Cutlass differ by more than 10%, the existing heuristic achieves 69% accuracy.  My new heuristic achieves 94% accuracy.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99513
* __->__ #99644



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Chillee committed:
tweak heuristic for sdpa selection based off of *data* (and a decision tree)

[ghstack-poisoned]

Chillee committed:
Update on "tweak heuristic for sdpa selection based off of *data* (and a decision tree)"

High level approach:
1. I generated a bunch of data comparing FlashAttention and Cutlass implementations (https://pastebin.com/pe0j3YeK) 
2. I trained a decision tree using standard train/val split methodology and hyperparameter sweeps (https://pastebin.com/fjYX1HjR).
2a. I did a bunch of feature augmentation to capture interactions between features.

The heuristic I ended up with is:
```
use_flash = seq_len / (num_heads * batch_size) > 6
```

TL;DR: On my dataset, where FlashAttention and Cutlass differ by more than 10%, the existing heuristic achieves 69% accuracy.  My new heuristic achieves 94% accuracy.




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Chillee committed:
Update on "tweak heuristic for sdpa selection based off of *data* (and a decision tree)"

High level approach:
1. I generated a bunch of data comparing FlashAttention and Cutlass implementations (https://pastebin.com/pe0j3YeK) 
2. I trained a decision tree using standard train/val split methodology and hyperparameter sweeps (https://pastebin.com/fjYX1HjR).
2a. I did a bunch of feature augmentation to capture interactions between features.

The heuristic I ended up with is:
```
use_flash = seq_len / (num_heads * batch_size) > 6
```

TL;DR: On my dataset, where FlashAttention and Cutlass differ by more than 10%, the existing heuristic achieves 69% accuracy.  My new heuristic achieves 94% accuracy.




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Chillee committed:
Update on "tweak heuristic for sdpa selection based off of *data* (and a decision tree)"

High level approach:
1. I generated a bunch of data comparing FlashAttention and Cutlass implementations (https://pastebin.com/pe0j3YeK) 
2. I trained a decision tree using standard train/val split methodology and hyperparameter sweeps (https://pastebin.com/fjYX1HjR).
2a. I did a bunch of feature augmentation to capture interactions between features.

The heuristic I ended up with is:
```
use_flash = seq_len / (num_heads * batch_size) > 6
```

TL;DR: On my dataset, where FlashAttention and Cutlass differ by more than 10%, the existing heuristic achieves 69% accuracy.  My new heuristic achieves 94% accuracy.




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -14034,7 +14034,7 @@
     CUDA: _scaled_dot_product_flash_attention_cuda
     NestedTensorCUDA: _scaled_dot_product_flash_attention_nestedtensor_cuda
 
-- func: _scaled_dot_product_flash_attention_backward(Tensor grad_out, Tensor query, Tensor key, Tensor value, Tensor out, Tensor logsumexp, Tensor cum_seq_q, Tensor cum_seq_k, int max_q, int max_k, float dropout_p, bool is_causal, int philox_seed, int philox_offse, *, float? scale=None) -> (Tensor grad_query, Tensor grad_key, Tensor grad_value)
```
thank you

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -39,22 +39,38 @@ struct sdp_params {
   if (params.query.is_nested() || params.key.is_nested() ||
       params.value.is_nested()) {
     // See check_for_nested_inputs for details
-    return {
```
this shortcut path can/should be removed with: https://github.com/pytorch/pytorch/pull/95438

However the `query_lengths{params.query.sym_size(2)}` will error with NTs. This seems like this will also be the case for the new heuristic but just a note to self really

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -70,13 +86,14 @@ inline std::array<SDPBackend, num_backends> priority_order(sdp_params params) {
+    // Revisit when updated cutlass kernel is upstreamed.
+    if (check_requires_grad(params, false)) {
+      if (6 * threads_flash > query_lengths) return efficient_first;
+    } else if ((small_threads_flash && more_threads_cutlass) || large_head_dim)
```
does the 94% accuracy use this combined heuristic or just `(6 * threads_flash > query_lengths)`?

Chillee replied:
The 94% accuracy is just with the 6 * threads_flash > query_lengths. I was only gathering data for fw + bw together (which is why there's a requires_grad guard).

@danthe3rd mentioned that the heuristic in xformers is only for the forwards pass, and in xformers they can use say, cutlass fw + flash bw (which I don't think we're able to do).

drisspg commented:
Epic! This is a great idea and I have been meaning to update the heuristic for some time now. I am planning on upstreaming the new version of Cutlass implementation next week and was going to wait for that. Do you think the pastebin scripts should be checked in somewhere? Since they are somewhat acting as the source of truth for perf eval of the different backends? We have other SDPA benchmarks here: https://github.com/pytorch/pytorch/blob/main/benchmarks/transformer/sdp.py.

Also haven't had a chance to do any testing on hopper but I assume this benchmark was done only on a100? Will want to re-run this see if the heuristic changes though as the implementations evolve

drisspg reviewed (approved):
Not sure about the other pt2 changes

Chillee commented:
@drisspg Yeah, happy to wait for the new cutlass impl if that's coming soon :)

Chillee committed:
Update on "tweak heuristic for sdpa selection based off of *data* (and a decision tree)"

High level approach:
1. I generated a bunch of data comparing FlashAttention and Cutlass implementations (https://pastebin.com/pe0j3YeK) 
2. I trained a decision tree using standard train/val split methodology and hyperparameter sweeps (https://pastebin.com/fjYX1HjR).
2a. I did a bunch of feature augmentation to capture interactions between features.

The heuristic I ended up with is:
```
use_flash = seq_len / (num_heads * batch_size) > 6
```

TL;DR: On my dataset, where FlashAttention and Cutlass differ by more than 10%, the existing heuristic achieves 69% accuracy.  My new heuristic achieves 94% accuracy.




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Chillee committed:
Update on "tweak heuristic for sdpa selection based off of *data* (and a decision tree)"

High level approach:
1. I generated a bunch of data comparing FlashAttention and Cutlass implementations (https://pastebin.com/pe0j3YeK) 
2. I trained a decision tree using standard train/val split methodology and hyperparameter sweeps (https://pastebin.com/fjYX1HjR).
2a. I did a bunch of feature augmentation to capture interactions between features.

The heuristic I ended up with is:
```
use_flash = seq_len / (num_heads * batch_size) > 6
```

TL;DR: On my dataset, where FlashAttention and Cutlass differ by more than 10%, the existing heuristic achieves 69% accuracy.  My new heuristic achieves 94% accuracy.




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Chillee commented:
@pytorchbot merge -f "failures unrelated"
